File: java2csharp/java/Microsoft.Mobile.Engagement/engagement/reach/EngagementReachContent.java
/*
 * Copyright (c) Microsoft Corporation. All rights reserved.
 */

package com.microsoft.azure.engagement.reach;

import static com.microsoft.azure.engagement.reach.ContentStorage.CATEGORY;
import static com.microsoft.azure.engagement.reach.ContentStorage.DLC;
import static com.microsoft.azure.engagement.reach.ContentStorage.DLC_ID;
import static com.microsoft.azure.engagement.reach.ContentStorage.PAYLOAD;
import static com.microsoft.azure.engagement.reach.ContentStorage.TTL;
import static com.microsoft.azure.engagement.reach.ContentStorage.USER_TIME_ZONE;

import java.util.TimeZone;

import org.json.JSONException;
import org.json.JSONObject;

import android.content.ContentValues;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;

/** Abstract class for reach content such as announcements and polls. */
public abstract class EngagementReachContent
{
  /** If set, the notification has downloadble content */
  public static final int FLAG_DLC_NOTIFICATION = 1;

  /** If set, the clickable content exists */
  public static final int FLAG_DLC_CONTENT = 2;

  /** Local identifier (which is a SQLite auto-increment) */
  private long mLocalID;

  /** DLC flags */
  private final int mDlc;

  /** DLC id */
  private final String mDlcId;

  /** Campaign id */
  final CampaignId mCampaignId;

  /** Category, null for default */
  final String mCategory;

  /** Body */
  String mBody;

  /** Expiry timestamp (ms since epoch), if any */
  private final Long mExpiry;

  /** Intent that must be launched for viewing this content */
  private Intent mIntent;

  /** True if replied, thus processed entirely, to avoid making that several times */
  private boolean mProcessed;

  /** Download identifier */
  private Long mDownloadId;

  /** Flag to remember DLC completes */
  private boolean mDlcCompleted;

  /**
   * Parse a campaign.
   * @param campaignId already parsed campaign id.
   * @param values content data.
   * @throws JSONException if payload parsing failure.
   */
  EngagementReachContent(CampaignId campaignId, ContentValues values) throws JSONException
  {
    /* Parse base fields */
    mCampaignId = campaignId;
    mDlc = values.getAsInteger(DLC);
    mDlcId = values.getAsString(DLC_ID);
    mCategory = values.getAsString(CATEGORY);
    Long expiry = values.getAsLong(TTL);
    if (expiry != null)
    {
      expiry *= 1000L;
      if (parseBoolean(values, USER_TIME_ZONE))
        expiry -= TimeZone.getDefault().getOffset(expiry);
    }
    mExpiry = expiry;
    if (values.containsKey(PAYLOAD))
      setPayload(new JSONObject(values.getAsString(PAYLOAD)));
  }

  /**
   * Parse boolean from content values.
   * @param values content values.
   * @param key key.
   * @return boolean value.
   */
  static boolean parseBoolean(ContentValues values, String key)
  {
    Integer val = values.getAsInteger(key);
    return val != null && val == 1;
  }

  /**
   * Get local identifier.
   * @return the local identifier.
   */
  public long getLocalId()
  {
    return mLocalID;
  }

  /**
   * Set the local identifier.
   * @param localId the local identifier.
   */
  void setLocalId(long localId)
  {
    /* Remember id */
    mLocalID = localId;
  }

  /** @return campaign id */
  public CampaignId getCampaignId()
  {
    return mCampaignId;
  }

  /** @return true if a download is necessary */
  public boolean hasDLC()
  {
    return mDlc > 0;
  }

  /** @return true if a download is necessary to display notification */
  public boolean hasNotificationDLC()
  {
    return (mDlc & FLAG_DLC_NOTIFICATION) == FLAG_DLC_NOTIFICATION;
  }

  /** @return true if a download is necessary after notification click */
  public boolean hasContentDLC()
  {
    return (mDlc & FLAG_DLC_CONTENT) == FLAG_DLC_CONTENT;
  }

  /** @return DLC id */
  public String getDlcId()
  {
    return mDlcId;
  }

  /** Set payload */
  void setPayload(JSONObject payload) throws JSONException
  {
    /* Read field */
    mDlcCompleted = true;
    mBody = payload.optString("body", null);
  }

  /**
   * Check if DLC completed.
   * @return true if DLC completed.
   */
  public boolean isDlcCompleted()
  {
    return mDlcCompleted;
  }

  /**
   * Get the category of this content. This category is also added in the intent that launches the
   * viewing activity if any. If this method returns <tt>null</tt>, the category list in the intent
   * will only contain <tt>android.intent.category.DEFAULT</tt>.
   * @return category.
   */
  public String getCategory()
  {
    return mCategory;
  }

  /**
   * Get content's body.
   * @return content's body.
   */
  public String getBody()
  {
    return mBody;
  }

  /**
   * Get the base intent to launch to view this content. The intent will be filtered with a
   * component name.
   * @return the base intent to launch to view this content.
   */
  public Intent getIntent()
  {
    if (mIntent == null && mDlcCompleted)
    {
      mIntent = buildIntent();
      if (mIntent != null)
        EngagementReachAgent.setContentIdExtra(mIntent, this);
    }
    return mIntent;
  }

  /**
   * Build the intent to launch to view this content.
   * @return the intent to launch to view this content.
   */
  abstract Intent buildIntent();

  /**
   * Get content expiration date (in ms since epoch).
   * @return content expiration date or null if not specified.
   */
  public Long getExpiry()
  {
    return mExpiry;
  }

  /**
   * Check if this content is now expired.
   * @return true iff this content is expired now.
   */
  public boolean hasExpired()
  {
    return mExpiry != null && System.currentTimeMillis() >= mExpiry;
  }

  /**
   * Drop content.
   * @param context any application context.
   */
  public void dropContent(Context context)
  {
    process(context, "dropped", null);
  }

  /**
   * Report content has been actioned.
   * @param context any application context.
   */
  public void actionContent(Context context)
  {
    process(context, "content-actioned", null);
  }

  /**
   * Report content been exited.
   * @param context any application context.
   */
  public void exitContent(Context context)
  {
    process(context, "content-exited", null);
  }

  /**
   * Check whether this content has a system notification.
   * @return true iff the content has a system notification.
   */
  public boolean isSystemNotification()
  {
    /* By default */
    return false;
  }

  /**
   * Dispose of this content so that new content can be notified. Possibly send feedback to the
   * service that sent it.
   * @param context application context.
   * @param status feedback status if any (null not to send anything).
   * @param extras extra information like poll answers.
   */
  void process(Context context, String status, Bundle extras)
  {
    /* Do it once */
    if (!mProcessed)
    {
      /* Send feedback if any */
      if (status != null)
        mCampaignId.sendFeedBack(context, status, extras);

      /* Mark this announcement as processed */
      mProcessed = true;

      /* Tell the reach application manager that announcement has been processed */
      EngagementReachAgent.getInstance(context).onContentProcessed(this);
    }
  }

  /**
   * Set content status.
   * @param values values from storage.
   */
  void setState(ContentValues values)
  {
    mDownloadId = values.getAsLong(ContentStorage.DOWNLOAD_ID);
  }

  /**
   * Get download identifier if a download has ever been scheduled for this content.
   * @return download identifier or null if download has never been scheduled.
   */
  public Long getDownloadId()
  {
    return mDownloadId;
  }

  /**
   * Set download identifier for this content.
   * @param context any application context.
   * @param downloadId download identifier.
   */
  public void setDownloadId(Context context, long downloadId)
  {
    mDownloadId = downloadId;
    EngagementReachAgent.getInstance(context).onDownloadScheduled(this, downloadId);
  }
}

File: java2csharp/csharp/Microsoft.Mobile.Engagement/engagement/reach/EngagementReachContent.cs
/*
 * Copyright (c) Microsoft Corporation. All rights reserved.
 */

namespace com.microsoft.azure.engagement.reach
{

//JAVA TO C# CONVERTER TODO TASK: This Java 'import static' statement cannot be converted to C#:
//	import static com.microsoft.azure.engagement.reach.ContentStorage.CATEGORY;
//JAVA TO C# CONVERTER TODO TASK: This Java 'import static' statement cannot be converted to C#:
//	import static com.microsoft.azure.engagement.reach.ContentStorage.DLC;
//JAVA TO C# CONVERTER TODO TASK: This Java 'import static' statement cannot be converted to C#:
//	import static com.microsoft.azure.engagement.reach.ContentStorage.DLC_ID;
//JAVA TO C# CONVERTER TODO TASK: This Java 'import static' statement cannot be converted to C#:
//	import static com.microsoft.azure.engagement.reach.ContentStorage.PAYLOAD;
//JAVA TO C# CONVERTER TODO TASK: This Java 'import static' statement cannot be converted to C#:
//	import static com.microsoft.azure.engagement.reach.ContentStorage.TTL;
//JAVA TO C# CONVERTER TODO TASK: This Java 'import static' statement cannot be converted to C#:
//	import static com.microsoft.azure.engagement.reach.ContentStorage.USER_TIME_ZONE;

	using JSONException = org.json.JSONException;
	using JSONObject = org.json.JSONObject;

	using ContentValues = android.content.ContentValues;
	using Context = android.content.Context;
	using Intent = android.content.Intent;
	using Bundle = android.os.Bundle;

	/// <summary>
	/// Abstract class for reach content such as announcements and polls. </summary>
	public abstract class EngagementReachContent
	{
	  /// <summary>
	  /// If set, the notification has downloadble content </summary>
	  public const int FLAG_DLC_NOTIFICATION = 1;

	  /// <summary>
	  /// If set, the clickable content exists </summary>
	  public const int FLAG_DLC_CONTENT = 2;

	  /// <summary>
	  /// Local identifier (which is a SQLite auto-increment) </summary>
	  private long mLocalID;

	  /// <summary>
	  /// DLC flags </summary>
	  private readonly int mDlc;

	  /// <summary>
	  /// DLC id </summary>
	  private readonly string mDlcId;

	  /// <summary>
	  /// Campaign id </summary>
	  internal readonly CampaignId mCampaignId;

	  /// <summary>
	  /// Category, null for default </summary>
	  internal readonly string mCategory;

	  /// <summary>
	  /// Body </summary>
	  internal string mBody;

	  /// <summary>
	  /// Expiry timestamp (ms since epoch), if any </summary>
	  private readonly long? mExpiry;

	  /// <summary>
	  /// Intent that must be launched for viewing this content </summary>
	  private Intent mIntent;

	  /// <summary>
	  /// True if replied, thus processed entirely, to avoid making that several times </summary>
	  private bool mProcessed;

	  /// <summary>
	  /// Download identifier </summary>
	  private long? mDownloadId;

	  /// <summary>
	  /// Flag to remember DLC completes </summary>
	  private bool mDlcCompleted;

	  /// <summary>
	  /// Parse a campaign. </summary>
	  /// <param name="campaignId"> already parsed campaign id. </param>
	  /// <param name="values"> content data. </param>
	  /// <exception cref="JSONException"> if payload parsing failure. </exception>
//JAVA TO C# CONVERTER WARNING: Method 'throws' clauses are not available in .NET:
//ORIGINAL LINE: EngagementReachContent(CampaignId campaignId, android.content.ContentValues values) throws org.json.JSONException
	  internal EngagementReachContent(CampaignId campaignId, ContentValues values)
	  {
		/* Parse base fields */
		mCampaignId = campaignId;
		mDlc = values.getAsInteger(DLC);
		mDlcId = values.getAsString(DLC_ID);
		mCategory = values.getAsString(CATEGORY);
		long? expiry = values.getAsLong(TTL);
		if (expiry != null)
		{
		  expiry *= 1000L;
		  if (parseBoolean(values, USER_TIME_ZONE))
		  {
			expiry -= TimeZone.Default.getOffset(expiry);
		  }
		}
		mExpiry = expiry;
		if (values.containsKey(PAYLOAD))
		{
		  Payload = new JSONObject(values.getAsString(PAYLOAD));
		}
	  }

	  /// <summary>
	  /// Parse boolean from content values. </summary>
	  /// <param name="values"> content values. </param>
	  /// <param name="key"> key. </param>
	  /// <returns> boolean value. </returns>
	  internal static bool parseBoolean(ContentValues values, string key)
	  {
		int? val = values.getAsInteger(key);
		return val != null && val == 1;
	  }

	  /// <summary>
	  /// Get local identifier. </summary>
	  /// <returns> the local identifier. </returns>
	  public virtual long LocalId
	  {
		  get
		  {
			return mLocalID;
		  }
		  set
		  {
			/* Remember id */
			mLocalID = value;
		  }
	  }


	  /// <returns> campaign id </returns>
	  public virtual CampaignId CampaignId
	  {
		  get
		  {
			return mCampaignId;
		  }
	  }

	  /// <returns> true if a download is necessary </returns>
	  public virtual bool hasDLC()
	  {
		return mDlc > 0;
	  }

	  /// <returns> true if a download is necessary to display notification </returns>
	  public virtual bool hasNotificationDLC()
	  {
		return (mDlc & FLAG_DLC_NOTIFICATION) == FLAG_DLC_NOTIFICATION;
	  }

	  /// <returns> true if a download is necessary after notification click </returns>
	  public virtual bool hasContentDLC()
	  {
		return (mDlc & FLAG_DLC_CONTENT) == FLAG_DLC_CONTENT;
	  }

	  /// <returns> DLC id </returns>
	  public virtual string DlcId
	  {
		  get
		  {
			return mDlcId;
		  }
	  }

	  /// <summary>
	  /// Set payload </summary>
//JAVA TO C# CONVERTER WARNING: Method 'throws' clauses are not available in .NET:
//ORIGINAL LINE: void setPayload(org.json.JSONObject payload) throws org.json.JSONException
	  internal virtual JSONObject Payload
	  {
		  set
		  {
			/* Read field */
			mDlcCompleted = true;
			mBody = value.optString("body", null);
		  }
	  }

	  /// <summary>
	  /// Check if DLC completed. </summary>
	  /// <returns> true if DLC completed. </returns>
	  public virtual bool DlcCompleted
	  {
		  get
		  {
			return mDlcCompleted;
		  }
	  }

	  /// <summary>
	  /// Get the category of this content. This category is also added in the intent that launches the
	  /// viewing activity if any. If this method returns <tt>null</tt>, the category list in the intent
	  /// will only contain <tt>android.intent.category.DEFAULT</tt>. </summary>
	  /// <returns> category. </returns>
	  public virtual string Category
	  {
		  get
		  {
			return mCategory;
		  }
	  }

	  /// <summary>
	  /// Get content's body. </summary>
	  /// <returns> content's body. </returns>
	  public virtual string Body
	  {
		  get
		  {
			return mBody;
		  }
	  }

	  /// <summary>
	  /// Get the base intent to launch to view this content. The intent will be filtered with a
	  /// component name. </summary>
	  /// <returns> the base intent to launch to view this content. </returns>
	  public virtual Intent Intent
	  {
		  get
		  {
			if (mIntent == null && mDlcCompleted)
			{
			  mIntent = buildIntent();
			  if (mIntent != null)
			  {
				EngagementReachAgent.setContentIdExtra(mIntent, this);
			  }
			}
			return mIntent;
		  }
	  }

	  /// <summary>
	  /// Build the intent to launch to view this content. </summary>
	  /// <returns> the intent to launch to view this content. </returns>
	  internal abstract Intent buildIntent();

	  /// <summary>
	  /// Get content expiration date (in ms since epoch). </summary>
	  /// <returns> content expiration date or null if not specified. </returns>
	  public virtual long? Expiry
	  {
		  get
		  {
			return mExpiry;
		  }
	  }

	  /// <summary>
	  /// Check if this content is now expired. </summary>
	  /// <returns> true iff this content is expired now. </returns>
	  public virtual bool hasExpired()
	  {
		return mExpiry != null && DateTimeHelperClass.CurrentUnixTimeMillis() >= mExpiry;
	  }

	  /// <summary>
	  /// Drop content. </summary>
	  /// <param name="context"> any application context. </param>
	  public virtual void dropContent(Context context)
	  {
		process(context, "dropped", null);
	  }

	  /// <summary>
	  /// Report content has been actioned. </summary>
	  /// <param name="context"> any application context. </param>
	  public virtual void actionContent(Context context)
	  {
		process(context, "content-actioned", null);
	  }

	  /// <summary>
	  /// Report content been exited. </summary>
	  /// <param name="context"> any application context. </param>
	  public virtual void exitContent(Context context)
	  {
		process(context, "content-exited", null);
	  }

	  /// <summary>
	  /// Check whether this content has a system notification. </summary>
	  /// <returns> true iff the content has a system notification. </returns>
	  public virtual bool SystemNotification
	  {
		  get
		  {
			/* By default */
			return false;
		  }
	  }

	  /// <summary>
	  /// Dispose of this content so that new content can be notified. Possibly send feedback to the
	  /// service that sent it. </summary>
	  /// <param name="context"> application context. </param>
	  /// <param name="status"> feedback status if any (null not to send anything). </param>
	  /// <param name="extras"> extra information like poll answers. </param>
	  internal virtual void process(Context context, string status, Bundle extras)
	  {
		/* Do it once */
		if (!mProcessed)
		{
		  /* Send feedback if any */
		  if (status != null)
		  {
			mCampaignId.sendFeedBack(context, status, extras);
		  }

		  /* Mark this announcement as processed */
		  mProcessed = true;

		  /* Tell the reach application manager that announcement has been processed */
		  EngagementReachAgent.getInstance(context).onContentProcessed(this);
		}
	  }

	  /// <summary>
	  /// Set content status. </summary>
	  /// <param name="values"> values from storage. </param>
	  internal virtual ContentValues State
	  {
		  set
		  {
			mDownloadId = value.getAsLong(ContentStorage.DOWNLOAD_ID);
		  }
	  }

	  /// <summary>
	  /// Get download identifier if a download has ever been scheduled for this content. </summary>
	  /// <returns> download identifier or null if download has never been scheduled. </returns>
	  public virtual long? DownloadId
	  {
		  get
		  {
			return mDownloadId;
		  }
	  }

	  /// <summary>
	  /// Set download identifier for this content. </summary>
	  /// <param name="context"> any application context. </param>
	  /// <param name="downloadId"> download identifier. </param>
	  public virtual void setDownloadId(Context context, long downloadId)
	  {
		mDownloadId = downloadId;
		EngagementReachAgent.getInstance(context).onDownloadScheduled(this, downloadId);
	  }
	}

}